src/restserver.c

```diff
@@ -455,24 +456,15 @@ int main(int argc, char *argv[])
         }
         rest_unlock(&rest);
 
-        res = select(FD_SETSIZE, &readfds, NULL, NULL, &tv);
-        if (res < 0)
+        res = ConnApi.f_step(rest.lwm2m, &tv);
```

Comment:
lwm2m/punica should be locked when calling lwm2m_handle_packet (which is done inside step). It would be better to separate connection and lwm2m layers. If possible, I think, that step function should be replaced with recvfrom-like function, which would return buffer and source connection. The lwm2m context could then be locked and lwm2m_handle packet called here.